Support for script dependencies

csrui opened the pull request:
Implements script dependencies through the use:

```
$enqueue->front([
    'as'     => 'ShirtCreatorInterface',
    'src'    => Helper::assetUrl('/js/app.js'),
    'dep'   => ['jquery'],
    'filter' => [ 'page' => '*' ]
]);
```

csrui committed:
Support for script dependencies

Implements script dependencies through the use:

```
$enqueue->front([
    'as'     => 'ShirtCreatorInterface',
    'src'    => Helper::assetUrl('/js/app.js'),
    'dep'   => ['jquery'],
    'filter' => [ 'page' => '*' ]
]);
```

ConnorVG commented:
This makes sense to me. Though, you should allow dep to be passed a string if there is only one of them, for sure :smile:

ConnorVG commented on Herbert/Framework/Enqueue.php:
```diff
@@ -66,13 +66,15 @@ public function buildInclude($attrs, $footer)
 //            $attrs['src'] = ltrim($attrs['src'], '/');
 //        }
 
+        if (!isset($attrs['dep'])) $attrs['dep'] = [];
```
I'd either new line this and use curlies or, on L77 use `array_get($attrs, 'dep', [])` instead.

ConnorVG commented on Herbert/Framework/Enqueue.php:
```diff
@@ -66,13 +66,15 @@ public function buildInclude($attrs, $footer)
         else
         {
-            wp_enqueue_script($attrs['as'], $attrs['src'], [], false, $footer);
+            wp_enqueue_script($attrs['as'], $attrs['src'], $attrs['dep'], false, $footer);
```
I'd definitely prepend the 3rd arg with `(array)` to allow for a single string dep to be added.

webstractions replied:
I would use 'deps' as the key just like WP Codex uses. 

Since enqueue-ing styles has dependancies, I would set `$attrs['deps']` before css check. 
`$attrs['deps'] = (array) array_get($attrs, 'dep', []);`

Now all is needed is:
`wp_enqueue_style($attrs['as'], $attrs['src'], $attrs['deps']);`
and
'wp_enqueue_script($attrs['as'], $attrs['src'], $attrs['deps'], false, $footer);'

csrui commented:
Most definitely a worthy improvement. Will take care of that.

csrui committed:
Update Enqueue.php

Improve dependency attributes fetching for enqueue.

csrui committed:
Adds support for script localizing

Changes the way scripts are called now using wp_register_script first in order to localize scripts.

Ex:

```
$enqueue->front([
    'as'     => 'MyScript',
    'src'    => Helper::assetUrl('/js/app.js'),
    'dep'   => 'jquery',
    'filter' => [ 'page' => '*' ],
    'localize' =>  array( 'ajaxurl' => admin_url( 'admin-ajax.php' ) )
]);
```

csrui commented:
Not sure if I should have created a new branch but since the source code affects the same area...

JacobDorman commented:
+1

PrafullaKumarSahu commented:
@csrui  I have just started using herbert and it makes me feel like a looser,as I think I am unable to do most of things like as you did, I tried doing 
```
$enqueue->front([
		    'as' => 'analytica',
			'src' => Helper::assetUrl('/jquery/analytica.js'),
			'dep' => ['jquery'],
			'filter' => [ 'page' => '*' ]
		]);
```

It gives me error  Call to a member function front() on a non-object in E:\plugins\Analytica\app\Controllers\AnalyticaController.php on line 166


ok, $enqueue is not an instance of some class, which class ?? I do not know :(

PrafullaKumarSahu commented:
@csrui  I have just started learning herbert from nearly a month and I do feel like a looser, as I am unable to do things usually I can do in wordpress, in last one month, using even your code 

```

public function __construct(Enqueue $enqueue){
        $this->enqueue = $enqueue;
    }


    $this->enqueue->front([
            'as' => 'analytica',
            'src' => Helper::assetUrl('/jquery/analytica.js'),
            'dep' => ['jquery'],
            'filter' => [ 'page' => '*' ]
        ]);






```

It gives me error  Fatal error: Using $this when not in object context in E:\plugins\Analytica\app\Controllers\AnalyticaController.php on line 168

and if I will call front with Enqueue::fronT() , it gives me no error neither it enqueue the script.

tormjens commented:
Hi @PrafullaKumarSahu 

You need can't use $this directly in a class without having it in a method.

Move your code up to the __construct-method, should work.

```
public function __construct(Enqueue $enqueue){
        $this->enqueue = $enqueue;

        $this->enqueue->front([
            'as' => 'analytica',
            'src' => Helper::assetUrl('/jquery/analytica.js'),
            'dep' => ['jquery'],
            'filter' => [ 'page' => '*' ]
        ]);
}
```

PrafullaKumarSahu commented:
Hi @tormjens  I have already solved this issue by en-queuing my scripts in enqueue.php
like below, any way thank you for your effort.

```
<?php namespace Analytica;

/** @var \Herbert\Framework\Enqueue $enqueue */

$enqueue->admin([
    'as'     => 'jquery',
    'src'    => 'jquery',
    'filter' => [ 'panel' => 'analytica-admin-settings' ]
]);

$enqueue->admin([
    'as'  => 'popup',
    'src' => Helper::assetUrl('/jquery/popup.js'), // => /resources/assets/js/popup.js
], 'footer');

$enqueue->admin([
    'as'  => 'analytica',
    'src' => Helper::assetUrl('/jquery/analytica.js'), // => /resources/assets/js/analytica.js
    'filter' => ['panel'=>'analytica-admin-settings']

], 'footer');

$enqueue->admin([
    'as'     => 'jquery-ui-datepicker',
    'src'    => 'jquery-ui-datepicker',
    'filter' => [ 'panel' => 'analytica-admin-settings' ]
], 'footer');

$enqueue->admin([
    'as'     => 'jquery-ui-datepicker',
    'src'    => 'http://ajax.googleapis.com/ajax/libs/jqueryui/1.8/themes/base/jquery-ui.css',
    'filter' => [ 'panel' => 'analytica-admin-settings' ]
]);

$enqueue->admin([
    'as'     => 'datepicker',
    'src'    => Helper::assetUrl('/jquery/datepicker.js'),
    'filter' => [ 'panel' => 'analytica-admin-settings' ]
], 'footer');

```

jeremyzahner commented:
@onnimonni This feature is crucial. Since the commits look firm, I see no reason for further delaying a merge.

jeremyzahner commented:
@csrui Could you resolve the merge conflicts? Thanks!

jeremyzahner commented:
@csrui Ping. =)

csrui commented:
This project seems dead.
